### PR TITLE
Add xml definition for service area cluster

### PR DESCRIPTION
**Pull request by hicklin**
Fixes https://github.com/project-chip/connectedhomeip/issues/33363.

This is the draft PR intended to facilitate development. It should not be merged.

### CODE REVIEW

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+  <bitmap name="Feature" type="bitmap32">
+    <cluster code="0x0150" />
+    <field name="ListOrder"       mask="0x01" />
+  </bitmap>
+
+  <!-- Enum definitions for structures-->
+  <enum name="OpStatusEnum" type="enum8">
+    <cluster code="0x0150"/>
+    <item value="0x00" name="Pending"/>
+    <item value="0x01" name="Traversing"/>
```
Comment (hicklin) [inline src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]: Where is this defined? This dose not seem to be part of the [service area spec PR](https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/8937/files).

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+  </struct>
+
+  <struct name="LocStruct">
+    <cluster code="0x0150"/>
+    <item name="LocationId"   type="int32u"                  optional="false" isNullable="false"/>
+    <item name="MapId"        type="int8u"                   optional="false" isNullable="true"/>
+    <item name="LocationInfo" type="LocationInfoStruct"      optional="false" isNullable="false"/>
+  </struct>
+
+  <struct name="ProgStruct">
```
Comment (hicklin) [inline src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]: Should we be a bit more verbose with the name of this structure like `ProgressStruct`?

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    <item value="0x02" name="DuplicatedLocations"/>
+    <item value="0x03" name="InvalidInMode"/>
+    <item value="0x04" name="InvalidSet"/>
+  </enum>
+
+  <!-- Enum definitions for skip location command response-->
+  <enum name="SkipCurrentStatus" type="enum8">
+    <cluster code="0x0150"/>
+    <item value="0x00" name="Success"/>
+    <item value="0x01" name="InvalidLocation"/>
```
Comment (hicklin) [inline src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]: In the spec, the name of this enum is `InvalidLocationList`. Should we keep to the spec?

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    <item value="0x02" name="InvalidInMode"/>
+  </enum>
+
+
+  <cluster>
+    <domain>General</domain>
+    <name>Service Area</name>
+      <code>0x0150</code>
+    <define>SERVICE_AREA_CLUSTER</define>
+    <description>The Service Area cluster provides an interface for controlling the locations where a device should operate, and for querying the current location.</description>
```
Comment (hicklin) [inline src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml:89]: Most clusters have the following lines before the description. Not sure what they do tbh. Have these been removed on purpose?

```
    <client init="false" tick="false">true</client>
    <server init="false" tick="false">true</server>
```

After the description there is usually a line defining the version of the cluster. If absent 1 will be assumed, which is correct for us, but it is nice to have it there for when the version needs to be changed.

```
<globalAttribute side="either" code="0xFFFD" value="1" />
```

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+    <!-- Attributes -->
+    <attribute side="server" code="0x0000" define="SupportedLocations"  type="array"     entryType="LocStruct"  writable="false" isNullable="false" optional="false">SupportedLocations</attribute>
+    <attribute side="server" code="0x0001" define="SupportedMaps"       type="array"     entryType="MapStruct"  writable="false" isNullable="false" optional="false">SupportedMaps</attribute>
```
Comment (hicklin) [inline src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]:
```suggestion
    <attribute side="server" code="0x0001" define="SupportedMaps"       type="array"     entryType="MapStruct"  writable="false" isNullable="true" optional="false">SupportedMaps</attribute>
```

According to [this PR](https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/8937).

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    <attribute side="server" code="0x0003" define="CurrentLocation"     type="int32u"                           writable="false" isNullable="true"  optional="true">CurrentLocation</attribute>
+    <attribute side="server" code="0x0004" define="EstimatedEndTime"    type="epoch_s"                          writable="false" isNullable="true"  optional="true" >EstimatedEndTime</attribute>
+    <attribute side="server" code="0x0005" define="Progress"            type="array"     entryType="ProgStruct" writable="false" isNullable="true"  optional="true">Progress</attribute>
+
+    <!-- Commands -->
+    <command source="client" code="0x00" name="SelectLocations" response="SelectLocationsResponse" optional="false">
+      <description>
+        Command used to select a set of device locations, where the device is to operate
+      </description>
+      <arg name="NewLocations" type="int32u" array="true"/>
```
Comment (hicklin) [inline src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml:216]: Should the type be `type="array" entryType="int32u"`?

Should this attribute be nullable `isNullable="true"`?

Comment (hicklin) [inline src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]:
```suggestion
      <arg name="NewLocations" type="int32u" array="true" isNullable="true"/>
```

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+    </command>
+
+    <command source="server" code="0x01" name="SelectLocationsResponse" disableDefaultResponse="true" optional="false">
+      <description>
+        This command is sent by the device on receipt of the SelectLocations command.
+      </description>
+      <arg name="Status"     type="SelectLocationsStatus"               optional="false"/>
```
Comment (hicklin) [inline src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]: I wonder if the type here has to be `enum8`, as defined in the spec. This depends on whether there is the possibility for us to want to add values form a different enum8, like vendor defined or a derived cluster.

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+      <description>
+        This command is used to skip the current location where the device operates.
+      </description>
+    </command>
+
+    <command source="server" code="0x03" name="SkipCurrentResponse" disableDefaultResponse="true" optional="true">
+      <description>
+        This command is sent by the device on receipt of the SelectLocations command.
+      </description>
+      <arg name="Status"     type="SkipCurrentStatus"              optional="false"/>
```
Comment (hicklin) [inline src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml:125]: Same comment as above about the type.

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+<configurator>
+  <domain name="CHIP"/>
+
+ <!-- semantic tag namespace definitons for service area-->
```
Comment (plauric) [inline src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]:
```suggestion
 <!-- semantic tag namespace definitions for service area-->
```

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml (hunk 1 above)
Comment (johnfierke) [reply src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]: It's in ServiceArea.adoc, near line 240.  The xml needs to be updated to remove Traversing, name change from OpStatusEnum to OperationalStatusEnum.

`[[ref_OperationalStatusEnum, OperationalStatusEnum]]
=== OperationalStatusEnum Type
This type defines the set of operational status values, and is derived from enum8. All values that are undefined SHALL be treated as reserved.
The following table defines the status values.

[options="header",valign="middle"]
|===
| Value   | Name       | Summary                                                                                                             | Conformance
| 0x00   s| Pending    | The device hasn't yet started operating, or hasn't finished but it's not currently operating, at the given location | M
| 0x01   s| Operating  | The device is currently operating at the given location                                                             | M
| 0x02   s| Skipped    | The device has skipped the given location, before or during operating at it                                         | M
| 0x03   s| Completed  | The device has completed operating at the given location  `

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml (hunk 2 above)
Comment (johnfierke) [reply src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]: The latest version of the spec PR has it as ProgressStruct.

Comment (plauric) [reply src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]: yes, this is called ProgressStruct in the spec

Comment (plauric) [inline src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]:
```suggestion
  <struct name="ProgressStruct">
```

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml (hunk 3 above)
Comment (johnfierke) [reply src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]: Please! - match the latest spec.

Comment (plauric) [inline src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]:
```suggestion
    <item value="0x01" name="InvalidLocationList"/>
```

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml (hunk 4 above)
Comment (johnfierke) [reply src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml:89]: service-area-cluster.xml was originally based on sample-mei-cluster.xml.

Since Service Area is going to be a standard cluster any additional lines that make sense should be added to the xml.

cluster version should definitely be added.

I'm not sure what the init lines do, they were not deliberately left out - just not in the sample xml.

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml (hunk 7 above)
Comment (johnfierke) [reply src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]: The spec says enum8, so unless we change the spec we should use that.

Should we change the spec?  (TL:DR - I don't know.)

The StatusText field does allow for vendor specific error information.
In general I'm in favor of strong use of types.

In operational-state-cluster.xml, which has derived clusters, the enum type is used for the type.
<attribute side="server" code="0x0004" define="OPERATIONAL_STATE"       type="OperationalStateEnum" 
However, the base server code deals with the state as a raw unsigned value.
uint8_t mOperationalState    (operational-state-server.h)

In the derived operational-state-rvc-cluster, additional states are defined, and the attribute is defined with the generic.
<attribute side="server" code="0x0004" define="OPERATIONAL_STATE"       type="enum8"

So there is at least one example where the base cluster definition uses the specific  type rather than the generic enum8 for an extensible value.  But it's not clear to me what the advantage is of the specific type in the base xml definition, since it isn't used in the server code. (Generated code for JSON conversion? Logging? - but then this isn't extended to derived classes.)

Comment (johnfierke) [reply src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]: There is also a GenericOperationalState defined in operational-state-cluster-objects.h which provides extensibility of the operational state values by associating a text label with specific enum values. 

These are stored in mOperationalStateList for specific derived implementations to provide definitions of all supported enum values - both standard and derived class specific - as in rvc-operational-state-delegate.h.

If we want to fully support derived classes having additional device specific state, we should probably add something similar.

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+        This command is used to skip the current location where the device operates.
+      </description>
+    </command>
+
+    <command source="server" code="0x03" name="SkipCurrentResponse" disableDefaultResponse="true" optional="true">
+      <description>
+        This command is sent by the device on receipt of the SelectLocations command.
+      </description>
+      <arg name="Status"     type="SkipCurrentStatus"              optional="false"/>
+      <arg name="StatusText" type="char_string"        length="64" optional="true"/>
```
Comment (johnfierke) [inline src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]: Latest spec has this length at 256

Comment (plauric) [inline src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]:
```suggestion
      <arg name="StatusText" type="char_string"        length="256" optional="true"/>
```

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+      </description>
+      <arg name="NewLocations" type="int32u" array="true"/>
+    </command>
+
+    <command source="server" code="0x01" name="SelectLocationsResponse" disableDefaultResponse="true" optional="false">
+      <description>
+        This command is sent by the device on receipt of the SelectLocations command.
+      </description>
+      <arg name="Status"     type="SelectLocationsStatus"               optional="false"/>
+      <arg name="StatusText" type="char_string"             length="64" optional="true"/>
```
Comment (plauric) [inline src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]:
```suggestion
      <arg name="StatusText" type="char_string"             length="256" optional="true"/>
```

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml (hunk 1 above)
Comment (hicklin) [reply src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]: Yeah, I was referring to the `Traversing` enum only. That's fine.

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml (hunk 4 above)
Comment (hicklin) [reply src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml:89]: OK. Lets leave the first two out and just add the version definition.

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml (hunk 7 above)
Comment (hicklin) [reply src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml]: I don't think we need to worry about supporting derived clusters at this stage except to consider if the way we're doing things will inhibit their addition in the future.

I also prefer to use strong types where possible. In this case, if we use the actual type, only that type will be accepted in the base implementation of the cluster. Hence, if this cluster is derived and if the derived version of this cluster would like to have derived cluster specific additions to this enum, they would be of a different type, as they would be in a different namespace, and therefor not accepted by the base cluster. To do this, we would also need to require that certain number spaces of the enum are reserved for derived clusters.

At this stage, the question we want to answer is, if we used the actual type, would we be able to change this to an enum8 in the future? I think that the answer to this question is yes.

If there is such a change, the SDK API of functions would change, requiring an enum8 instead of the actual type. This change would be something that needs to be implemented by the vendors that use this cluster when they choose to use the later version of the SDK. However, since this is not a change to the data on the wire (the data being sent back and fort would still be a uint8), we can argue that this would not be a breaking change as clients and devices using the different versions of the SDK would still be able to communicate.

## TLDR

Given that currently we do not see the future need of clusters derived form this one, we can go ahead and use the actual types as changing this to an enum8 in the future is not a breaking change.

File: src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+      <item value="0x16" name="Tile"/>
+      <item value="0x17" name="Vinyl"/>
+    </enum>
+
+
+  <bitmap name="Feature" type="bitmap32">
+    <cluster code="0x0150" />
+    <field name="ListOrder"           mask="0x01" />
+    <field name="SelectWhileRunning"  mask="0x02" />
+  </bitmap>
```
Comment (johnfierke) [inline src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml:24]: Feature definition is apparently much richer than whatever example I based this on, also looks like it should be moved under the "cluster" section.  (Might be due to the newer zap version than I originally used?)

From  microwave-oven-control-cluster.xml

`features>  
      <feature bit="0" code="PWRNUM" name="PowerAsNumber" summary="Power is specified as a unitless number or a percentage">
        <optionalConform choice="a"/>
      </feature>
      <feature bit="1" code="WATTS" name="PowerInWatts" summary="Power is specified in Watts">
        <optionalConform choice="a"/>
      </feature>
      <feature bit="2" code="PWRLMTS" name="PowerNumberLimits" summary="Supports the limit attributes used with the PWRNUM feature">
        <optionalConform>
          <feature name="PWRNUM"/>
        </optionalConform>
      </feature>
    </features>`

Comment (hicklin) [reply src/app/zap-templates/zcl/data-model/chip/service-area-cluster.xml:24]: Moved to CHIP PR.

**Comment by hicklin** [comment]
Closing as this PR was moved to CHIP.
